### ISSUE TITLE
test-leaks seems to be flaky / "Some object was not garbage collected!"

### ISSUE THREAD

**Issue by blueyed** [awesomeWM/awesome#808]
Build: https://travis-ci.org/awesomeWM/awesome/jobs/120752692

```
== Running /home/travis/build/awesomeWM/awesome/tests/test-leaks.lua ==
+start_awesome
+export DISPLAY=:5
+cd /home/travis/build/awesomeWM/awesome/tests/../build
+awesome_pid=13416+DISPLAY=:5 /home/travis/build/awesomeWM/awesome/tests/../build/awesome -c /home/travis/build/awesomeWM/awesome/tests/../build/awesomerc.lua
+cd -
+wait_until_success wait for awesome startup via awesome-client echo 'return 1' | DISPLAY=:5 '/home/travis/build/awesomeWM/awesome/tests/../utils/awesome-client' 2>&1
+[ true = true ]
+set +x
+[ ! -r /home/travis/build/awesomeWM/awesome/tests/test-leaks.lua ]
+DISPLAY=:5 /home/travis/build/awesomeWM/awesome/tests/../utils/awesome-client
+cat /home/travis/build/awesomeWM/awesome/tests/test-leaks.lua
+tail -n 100000 -f --pid 13416 /tmp/tmp.tirTf7WQk1/_awesome_test.log
2016-04-04 22:36:06 W: Cannot reliably detect EOF on an GIOInputStream with this LGI version
Some object was not garbage collected!
2016-04-04 22:36:07 W: awesome: luaA_dofunction:77: error while running function!
stack traceback:
	[string "-- Some memory leak checks as integration tes..."]:38: in function 'collectable'
	[string "-- Some memory leak checks as integration tes..."]:66: in function 'f'
	lib/awful/remote.lua:25: in function <lib/awful/remote.lua:21>
error: @lib.wibox.widget.imagebox (table: 0x41002b90)
Error: run_steps() was never called
+grep -q -E ^Test finished successfully$ /tmp/tmp.tirTf7WQk1/_awesome_test.log
+echo ===> ERROR running /home/travis/build/awesomeWM/awesome/tests/test-leaks.lua! <===
===> ERROR running /home/travis/build/awesomeWM/awesome/tests/test-leaks.lua! <===
+grep --color -o --binary-files=text -E .*[Ee]rror.*|.*assertion failed.* /tmp/tmp.tirTf7WQk1/_awesome_test.log
2016-04-04 22:36:07 W: awesome: luaA_dofunction:77: error while running function!
error: @lib.wibox.widget.imagebox (table: 0x41002b90)
Error: run_steps() was never called
```

I've read some comments regarding this issue somewhere (IIRC) - related to
slow/fast Travis builds, but we should have an issue to track this.

Do you think wrapping the tests with `travis-retry` to retry them 3 times is a good idea?

**Comment by Elv13** [comment]
So far it has been quite random, but always in the luajit target. I guess re-trying would help, if only to avoid having a `[build|failure]` tag in the README.md

**Comment by psychon** [comment]
Hm... We could do a travis build that only runs on LuaJIT, only runs this one test and does e.g. 100 times. That way we would get some rough estimate on how often it happens. Then we can replace the three GC cycles with 100 times and see if that helps. Perhaps adding more `emit_refresh()` calls to `test-leaks.lua`?

Besides that I don't have a good idea on what to try. The code in `lib/awful/widget/layoutbox.lua` is quite simple and really should not leak.

**Comment by blueyed** [comment]
Ubuntu Trusty on Travis has luajit 2.0.2 (http://packages.ubuntu.com/trusty/luajit), and the current release is 2.0.4.
Maybe it's an issue with them after all?
I am trying to build luajit from source in https://github.com/awesomeWM/awesome/pull/812.

**Comment by blueyed** [comment]
luajit 2.0.2 is from 2013-06-03 after all..!
#812 seems to be ready - it has not failed yet.. ;)

**Comment by psychon** [comment]
Thanks to @Elv13, I had a new idea on how to investigate. I have a Lua library that traverses all reachable objects and prints how each thread (coroutine) is reachable (because these are easy to identify and there aren't many of them). The library is https://gist.github.com/psychon/2797b81f23520d2285c44a28c85cd017 and can be used like `require("traverse").new():do_all()` (originally I wanted to measure memory usage, hence this weird usage).

Anyway, I changed the code in layoutbox.lua to add `w._thread = coroutine.create(function() end)` and let `test-leaks.lua` run against LuaJIT for a while (since I still cannot reproduce with "plain" Lua).

The result is (ignoring the normal, expected entries that Lua creates internally):
```
registry[awesome.object.registry][userdata: 0x40bb1890](env)[userdata: 0x40f08588]f:lib/wibox/drawable.lua:368(up:ret)[widget][third][widgets][4][_thread]
```
Now, what does this mean? I'm not completely sure.

`drawable.lua:368` is some "internal magic" to make `button::press` and `button::release` work on widgets, based on the same signals on a drawin/wibox. This function references the drawable (upvalue `ret`), which has an entry `.widget.third.widgets[4]._thread`, so I guess this is a "normal" wibox which contains an align layout, containing a fixed or flex layout whose fourth entry is the layoutbox.

So, what is really leaked here seems to be a wibox, but for some reason the other tests do not catch this (or the layoutbox is just the first singleton that is being tested?).

Now, why is the wibox referenced. Being reachable through `debug.getregistry()["awesome.object.registry"]` means that the C code has a reference to some object (`luaA_object_ref()` / `luaA_object_unref()`). Sadly, this is done as a lightuserdata and so we only see the pointer and not what type it really is. Having something reachable through `(env)` here means that `luaA_object_ref_item()` was used to make the object refer to some other object. In this case, this likely means something related to connecting the signal callback in `drawable.lua`.

My guess is:

TL;DR: Some wibox is still visible and thus cannot be garbage collected and keeps all its widgets alive.

**Comment by psychon** [comment]
Uhmmm... New guess: The above is due to the mouse pointer being inside of the wibox. The C code will then create an extra reference to the drawin so that it can latter emit `mouse::leave` on it (and the item being referenced surely is the drawin's drawable).

Thus, I guess I still cannot reproduce this issue.... either that, or this really is the issue we have been looking for for so long.

**Comment by psychon** [comment]
The following patch "fixes" the problem (couldn't reproduce the problems in 1000 attempts with this patch, while without it it's trivially reproducible):
```patch
diff --git a/lib/awful/tooltip.lua b/lib/awful/tooltip.lua
index 6c37aea..82678e3 100644
--- a/lib/awful/tooltip.lua
+++ b/lib/awful/tooltip.lua
@@ -274,6 +274,9 @@ tooltip.new = function(args)
         end
     end
 
+    -- Force the wibox to be created TODO: I don't want this :-(
+    instance_mt.__index(self, "wibox")
+
     return self
 end
 
```
I have zero clue on what is going on here.

**Comment by psychon** [comment]
How about the following for "fixing" this:
```diff
diff --git a/tests/test-leaks.lua b/tests/test-leaks.lua
index 247524f..031c388 100644
--- a/tests/test-leaks.lua
+++ b/tests/test-leaks.lua
@@ -63,7 +63,9 @@ collectable(wibox.layout.align.horizontal())
 collectable(awful.widget.launcher({ image = cairo.ImageSurface(cairo.Format.ARGB32, 20, 20), command = "bash" }))
 collectable(awful.widget.prompt())
 collectable(awful.widget.textclock())
-collectable(awful.widget.layoutbox(1))
+
+-- FIXME: This randomly fails (and more often so on LuaJIT?!?)
+--collectable(awful.widget.layoutbox(1))
 
 -- Some widgets do things via timer.delayed_call
 prepare_for_collect = emit_refresh
```
Do other test cases in their need similar treatment? I think I remember something about tooltips leaking, too....

**Comment by Elv13** [comment]
I got a new "fix" here
https://github.com/awesomeWM/awesome/pull/878/commits/5e588486910b915f1708d430e66b3a1527c24f29

Still ugly, but cleaner than my last one. It happen with all Lua versions, see
https://travis-ci.org/awesomeWM/awesome/jobs/128603698#L3573 for Lua 5.3. It's actually quite common in 5.1 and even more in luajit.
